tests/models/ranking/test_candidate_ranking.py

```diff
@@ -0,0 +1,291 @@
+            [
+                [70, 11, 1, "2021-11-30"],
+                [70, 12, 1, "2021-11-30"],
+                [10, 11, 1, "2021-11-30"],
+                [10, 12, 1, "2021-11-29"],
+                [10, 13, 9, "2021-11-28"],
+                [20, 11, 1, "2021-11-27"],
+                [20, 14, 2, "2021-11-26"],
+                [30, 11, 1, "2021-11-24"],
+                [30, 12, 1, "2021-11-23"],
+                [30, 14, 1, "2021-11-23"],
+                [30, 15, 5, "2021-11-21"],
+                [40, 11, 1, "2021-11-20"],
+                [40, 12, 1, "2021-11-19"],
+            ],
+            columns=Columns.Interactions,
+        )
+        user_id_map = IdMap.from_values([10, 20, 30, 40, 50, 60, 70, 80])
+        item_id_map = IdMap.from_values([11, 12, 13, 14, 15, 16])
+        interactions = Interactions.from_raw(interactions_df, user_id_map, item_id_map)
+        return Dataset(user_id_map, item_id_map, interactions)
+
+    @pytest.fixture
+    def users(self) -> tp.List[int]:
+        return [10, 20, 30]
+
+    @pytest.fixture
+    def model(self) -> PopularModel:
+        return PopularModel()
+
+    def test_get_train_with_targets_for_reranker_happy_path(self, model: PopularModel, dataset: Dataset) -> None:
+        candidate_generators = [CandidateGenerator(model, 2, False, False)]
+        splitter = TimeRangeSplitter("1D", n_splits=1)
+        sampler = PerUserNegativeSampler(1, 32)
+        two_stage_model = CandidateRankingModel(
+            candidate_generators,
+            splitter,
+            sampler=sampler,
+            reranker=CatBoostReranker(CatBoostRanker(random_state=32, verbose=False)),
+        )
+        actual = two_stage_model.get_train_with_targets_for_reranker(dataset)
+        expected = pd.DataFrame(
+            {
+                Columns.User: [10, 10],
+                Columns.Item: [14, 11],
+                Columns.Target: np.array([0, 1], dtype="int32"),
+            }
+        )
+        pd.testing.assert_frame_equal(actual, expected)
+
+    def test_recommend_happy_path(self, model: PopularModel, dataset: Dataset) -> None:
+        candidate_generators = [CandidateGenerator(model, 2, True, True)]
+        splitter = TimeRangeSplitter("1D", n_splits=1)
+        sampler = PerUserNegativeSampler(1, 32)
+        two_stage_model = CandidateRankingModel(
+            candidate_generators,
+            splitter,
+            sampler=sampler,
+            reranker=CatBoostReranker(CatBoostRanker(random_state=32, verbose=False)),
+        )
+        two_stage_model.fit(dataset)
+
+        actual = two_stage_model.recommend(
+            [10, 20, 30],
+            dataset,
+            k=3,
+            filter_viewed=True,
+        )
+        expected = pd.DataFrame(
+            {
+                Columns.User: [10, 10, 20, 20, 30],
+                Columns.Item: [14, 15, 12, 13, 13],
+                Columns.Score: [
+                    -0.1924785067861987,
```

Comment:
with `atol=0.001` there is no need to specify quadrillion decimal digits

Reply:
left three digits after the decimal point

**done**